cypress-io/cypress
# feat(app): run ct and e2e specs in unified runner app

There is a bit going on here, here is my ideas on how to move forward w/ building out the unified runner. Probably best to pair review/go over this together. This will at least unblock working on the UI for the rest of the team, even if the architecture is not perfect.

## How to use/test

1. `yarn dev`
2. Choose app or launchpad
3. Choose e2e or CT
4. Launch browser
5. Navigate to `__vite__/` route
6. Should show new unified runner, and be able to execute specs in the new runner
7. UI is ugly, focus is the model and functionality right now

Here is how it looks right now, hoping to merge this up and point someone at the UI soon:

![image](https://user-images.githubusercontent.com/19196536/137657674-e978bfb9-71b3-4fb4-8db8-586583bbf088.png)

**Note** you can avoid navigating to `__vite__/` manually by launching with `UNIFIED_RUNNER=1`.


## Unified components

We have some duplication between runner and runner-ct. Most of the common code is in runner-shared, though. The new runner app uses Vite and ES modules. We can't seamlessly import a lot of code from `driver`, `reporter` or `runner-shared` into the unified runner due to circular imports and dependencies (webpack goes off the ES modules spec to resolve these, but Vite does not).

For this reason, the way we access `driver`, `reporter` etc is by building a bundle that imports them using webpack, then injecting the bundle using a `<script>` tag. Everything is placed on `window.UnifiedRunner`. This also includes dependencies like MobX, singletons like `eventManager`, and components like `Reporter`. So you can do `window.UnifiedRunner.Reporter`, `window.UnifiedRunner.eventManager` to access those modules. The entrypoint for this bundle is `packages/runner-ct/src/main.tsx`, so look there for full list of things on `window.UnifiedRunner`. 

## Type all the things

As we go, I think we should move code from `runner-shared` to `app`. I did this for some critical parts of the system, like `iframe-model` and `AutIframe`. We can add type definitions as we go. Since anything in `app` uses Vite, it cannot import things from `runner-shared` (circular deps issue), so instead we just pass the dependencies using dependency injection. Example:

Before:

```
// runner-shared/src/aut-iframe.js
import eventManager from './eventManager'

class AutIframe {
  constructor () { 
    eventManager.listen()
  }
}
```

After:

```
// runner-shared/src/aut-iframe.ts

class AutIframe {
  // NOTE: do NOT import `eventManager`  since it comes from runner-shared and attempts
  // to import driver, which has circular deps that Vite cannot resolve
  // instead, inject via the constructor
  constructor (private eventManager: any) { 
    this.eventManager.listen()
  }
}

/* Usage */
new AutIframe(window.UnifiedRunner.eventManager) // <- inject dependency 
```

**note** we now pass anything that would be imported from `runner-shared` as an argument

The eventual goal is all the code is typed and lives in `app`. This is practically impossible without a significant rejigging of `driver` and `reporter`, but we can make sure any new components or front-end logic lives here and is fully typed.

There are some outstanding questions:

- how to we manage the current spec? Previously this was with the URL, but now we are using GraphQL as a source of truth. I set it up in this manner, but it feels slower (GraphQL adds latency, maybe I'm doing it wrong)
- figure out how run mode should work (CT and E2E are slightly different as it stands)
- probably a lot more, to discuss w/ @brian-mann
